### ISSUE TITLE
all request_headers are filtered out by default

### ISSUE THREAD

**Issue by PragTob** [appsignal/appsignal-elixir#649]
Hello friends! Long time no see 👋 

## Describe the bug

We're getting no HTTP header information in our bug traces, not even the [defaults](https://docs.appsignal.com/elixir/configuration/options.html#option-request_headers) which are specified as `["accept", "accept-charset", "accept-encoding", "accept-language", "cache-control", "connection", "content-length", "path-info", "range", "request-method", "request-uri", "server-name", "server-port", "server-protocol"]`

We haven't configured the http_headers anywhere (full text search for appsignal on the App and checked our environment variables)

## To Reproduce

- Using AppSignal for Elixir package version 2.1.6
- used with phoenix and the phoenix integration
- do not configure request headers anywhere
- tigger error notification
- no request headers 😭 

![Screenshot 2021-05-11 at 17 09 56](https://user-images.githubusercontent.com/606517/117839734-c756ad00-b27b-11eb-8f1a-e3f9b9cd1efb.png)

## Diagnosis

I believe this was introduced by #637 making the default return value of request_headers an empty list because as the [docs say](https://docs.appsignal.com/elixir/configuration/options.html#option-request_headers):

>  To configure AppSignal to not store any HTTP request headers on AppSignal transactions, configure the option with an empty list.

## Ask

Please let me know if the default will be restored or if this change is intended in which case I'll make it explicit in our app and an update to the docs would be great 😁  Or, let me know where I went wrong 👀 

-------------------------

Anyhow, 
thank you all! 💚 

![IMG_20180525_145837](https://user-images.githubusercontent.com/606517/117839424-81014e00-b27b-11eb-9eba-45948066f133.jpg)

**Comment by tombruijn** [comment]
Hi @PragTob, thanks for the report! I'm going to try and reproduce your scenario and will you let you know if I need any more information or know more about the issue.

**Comment by tombruijn** [comment]
@PragTob I found something.

The PR #637 you linked is I think only for when AppSignal isn't active and the config isn't set, to prevent an error from occurring. By default it uses [this list from the default config Map](https://github.com/appsignal/appsignal-elixir/blob/cd9ef41d5957ed68b4a8f75a498fde082e82f0f7/lib/appsignal/config.ex#L27-L31). I don't think that change is the cause of this issue.

With the config option `request_headers: []`, the headers are not listed reported at all. Instead they are reported as "[FILTERED]" here. Other metadata like "host" and "method" (which are not headers) are also replaced with "[FILTERED]".

Which means they are being filtered out by our filtering system. I suspect the `filter_data_keys` config option is configured. I get this behavior with this config:

```elixir
config :appsignal, :config, filter_data_keys: [""]
```

Is the `filter_data_keys` config option configured in your app? If so, what values does it contain?
You can also use our [diagnose tool](https://docs.appsignal.com/elixir/command-line/diagnose.html) to see what config AppSignal picks up.

**Comment by PragTob** [comment]
@tombruijn thanks!

Weird... nothing like that is configured to the best of my knowledge.

I ran the daignostics locally in MIX_ENV=prod:

```
Configuration
  active: false
  ca_file_path: "path"
  debug: false
  diagnose_endpoint: "https://appsignal.com/diag"
  dns_servers: []
  enable_host_metrics: true
  enable_minutely_probes: true
  endpoint: "https://push.appsignal.com"
  env: :dev
  files_world_accessible: true
  filter_data_keys: []
  filter_parameters: []
  filter_session_data: []
  ignore_actions: ["_something_"]
    Sources:
      default: []
      file:    ["_something_"]
  ignore_errors: []
  ignore_namespaces: []
  log: "file"
  otp_app: :tiger (Loaded from file)
  request_headers: ["accept", "accept-charset", "accept-encoding", "accept-language", "cache-control", "connection", "content-length", "path-info", "range", "request-method", "request-uri", "server-name", "server-port", "server-protocol"]
  revision: nil (Loaded from file)
  send_params: true
  skip_session_data: true
    Sources:
      default: false
      file:    true
  transaction_debug_mode: false
  valid: false
```

we do:

```
config :phoenix,
       :filter_parameters,
       {:keep,
        [stuff]
````

that shouldn't take it away though as the data shows up in our logs.

I did check the env variables and the only APPSIGNAL ones are the key etc so no config options 🤔

**Comment by tombruijn** [comment]
@PragTob I think I know what's happening, I'm just not sure why yet exactly, but this is what I know so far.

In the AppSignal Elixir integration we [merge all the filter config options](https://github.com/appsignal/appsignal-elixir/blob/cd9ef41d5957ed68b4a8f75a498fde082e82f0f7/lib/appsignal/config.ex#L57-L61) into the `filter_data_keys` config option: filter_parameters, filter_sesison_data, but also the Phoenix filter_parameters config option.

When you set a "keep" value we don't parse the "keep" as part of an allow list. The filter_data_keys config option is only used as a deny list. The "keep" value is dropped and filter_parameters list is now configured as an empty list.

Now on the Rust side we expect a comma separate string, but it gets an empty string. Our extension then tries to parse that config, which it can't, and sets `None` (a Rust value) instead of `Some(list,of,keys)`.
But from what I can tell None should be interpreted as "filter none of the params" rather than "filter everything" 🤔

I think there are two things happening:

- The filter_parameters config option with an empty list is maybe interpreted wrong.
- We don't support `keep` in Phoenix filter_parameters, like so:
  - `config :phoenix, :filter_parameters, {:keep, ["param-name"]}`

For now I can recommend to not use the `keep` option in the filter_parameters option and explicitly filter out params. We'll have a look at the issue with the team and see how we can best fix it.

**Comment by tombruijn** [comment]
Looks like we should support Phoenix's keep option: https://github.com/appsignal/appsignal-elixir/pull/563
More research required.

**Comment by tombruijn** [comment]
I looked a little further with tips from Jeff and I see that all the request headers and other data is being filtered when the phoenix filter_parameters is used in [MapFilter](https://github.com/appsignal/appsignal-elixir/blob/cd9ef41d5957ed68b4a8f75a498fde082e82f0f7/lib/appsignal/utils/map_filter.ex#L9-L13)

```
config :phoenix, :filter_parameters, {:keep, ["param-name"]
```

So it's not an extension issue.

The integration just uses one config option for the filtering keys, so it filters out all data from other data object that contain params, headers, session data, etc.

If we don't want this to happen we should probably split that config option or only filter params using the Phoenix filter_parameters config option.

**Comment by tombruijn** [comment]
@PragTob We've discussed it in our team meeting and we realized we need to change how this system works. We're going to split the different config options for filters out again into separate filters, and add a allowlist system in the extension as well (rather than implement that in the Elixir integration as we do now). To fix this we will need to rewrite part of this system, which will take a while. We don't know exactly when we'll pick it up, but it's on our list now.

Until then the only way not to filter out everything is to add entries for all the things you do want to keep to the phoenix allow list.

**Comment by PragTob** [comment]
HI @tombruijn thanks for the thorough investigation! I'm a bit confused now though - so is the state of the art still that `:keep` is ignored/treated as an empty list and the suggestion would be to not use `keep`?

That'd be a bit inconvenient, I generally prefer allow lists over deny lists as they are much less prone to accidentally leaking confidential data.

Thanks a bunch either way!

![IMG_20180818_152929](https://user-images.githubusercontent.com/606517/118612851-79333380-b7be-11eb-98a3-e6ef2548ca2d.jpg)

**Comment by tombruijn** [comment]
@PragTob You can use the `keep`, but you need specify all the metadata (request headers, query params and session data) using the Phoenix `keep` parameters option. You need to specify more than query params with that option.

It's a bit weird, but that's how the filtering system would currently filter it out otherwise. An example with a bit of everything:

```elixir
config :phoenix, :filter_parameters, {:keep, ["query-param-name", "req_headers.accept", "host", "method"]}
```

**Comment by PragTob** [comment]
that is... weird yes 😅  Thanks for the input I'll see if I can make this work.

**Comment by PragTob** [comment]
@tombruijn hey I deployed this as a work around now and it works - thanks! Shall we close this or keep it open for some follow up work on this (docs or change the behaviour or what not) or is this considered working as intended now?

**Comment by tombruijn** [comment]
@PragTob great to hear the workaround works! Let's keep this issue open until we have planned work on the project ([private issue](https://github.com/appsignal/integration-guide/issues/31)) needed to change how this works.

**Comment by tombruijn** [comment]
@PragTob A fix for this issue has been released in AppSignal for Elixir package 2.2.8.
